### PR TITLE
Raspberry Pico 2 porting

### PR DESCRIPTION
**Pull request by PeterBee97**
## Summary
Porting of RP2350 platform continuing @casaroli 's work #13305. Still have many stuff to test and fix, but most things are working!

## Impact
* drivers:
  
  * [x]  timerisr - using sysclock -- need to switch to TICK
  * [x]  UART
  * [x]  led/gpio/button/ws2812
  * [x]  usbdev serial -- somewhat
  * [x]  uniqueid
  * [ ]  PSRAM
  * [x]  ADC
  * [x]  Flash boot
  * [x]  I2C
  * [ ]  usbmsc
  * [ ]  usb composite
  * [ ]  AON timer (RTC)
  * [x]  watchdog
  * [x]  I2S
  * [x]  SPI
  * [x]  spidev
  * [x]  pio
  * [x]  SMP
  * [ ]  MTD
* boards
  
  * [x]  Raspberry Pi Pico 2
  * [ ]  Pimoroni tiny2350
  * [ ]  Pimonori Pico 2 Plus
* chore
  
  * [x]  register definition generate / reformat
  * [x]  format all source files

## Impact
New potentially very popular boards supported in NuttX.
 
## Testing
Manual testing with the boards on hand.

### CODE REVIEW

**Comment by PeterBee97** [comment]
The PSRAM part is removed for now since I don't have a Pimoroni board at hand to test. Calling for help :)

**Comment by hartmannathan** [comment]
> The PSRAM part is removed for now since I don't have a Pimoroni board at hand to test. Calling for help :)

@PeterBee97 you could try writing to the mailing list and asking if someone has the board.

File: Documentation/platforms/arm/rp23xx/boards/raspberrypi-pico-2/README.txt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+4. Connect Raspberry Pi Pico 2 board to USB port while pressing BOOTSEL.
+   The board will be detected as USB Mass Storage Device.
+   Then copy "nuttx.uf2" into the device.
+   (Same manner as the standard Pico SDK applications installation.)
+
+5. To access the console, GPIO 0 and 1 pins must be connected to the
+   device such as USB-serial converter.
+
+   `usbnsh` configuration provides the console access by USB CDC/ACM serial
+   devcice.  The console is available by using a terminal software on the USB
```
Comment (hartmannathan) [inline Documentation/platforms/arm/rp23xx/boards/raspberrypi-pico-2/README.txt:38]:
```suggestion
   device.  The console is available by using a terminal software on the USB
```

File: Documentation/platforms/arm/rp23xx/index.rst [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+using XIP. By using the default `BOOT_RUNFROMFLASH` configuration,
+the full image is run from the flash making most of the internal
+SRAM available for the OS and applications, however the execution
+is slower. The cache can speed up, but you might want set your
+time critical functions to be placed in the SRAM (copied from
+the flash on startup).
+
+It is also possible to execute from SRAM, which reduces the
+available SRAM to the OS and applications, however it is very
+useful when debugging as erasings and rewriting the flash on
```
Comment (hartmannathan) [inline Documentation/platforms/arm/rp23xx/index.rst:97]:
```suggestion
useful when debugging as erasing and rewriting the flash on
```

File: Documentation/platforms/arm/rp23xx/index.rst [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+every build is tedious and slow. This option is enabled with
+`BOOT_RUNFROMISRAM` and requires `openocd`` and/or `gdb`.
+
+There is a third option which is to write the firmware on the
+flash and it gets copied to the SRAM. This is enabled with
+`CONFIG_BOOT_COPYTORAM` and might be useful for time critical
+applications, on the expense of reduced useable interna SRAM
```
Comment (hartmannathan) [inline Documentation/platforms/arm/rp23xx/index.rst:104]:
```suggestion
applications, on the expense of reduced usable internal SRAM
```

File: Documentation/platforms/arm/rp23xx/index.rst [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+module and companion software (openocd and gdb)
+
+Running NuttX
+=============
+
+Most builds provide access to the console via UART0.  To access this
+GPIO 0 and 1 pins must be connected to the device such as USB-serial converter.
+
+The `usbnsh` configuration provides the console access by USB CDC/ACM serial
+devcice.  The console is available by using a terminal software on the USB host.
```
Comment (hartmannathan) [inline Documentation/platforms/arm/rp23xx/index.rst:198]:
```suggestion
device.  The console is available by using a terminal software on the USB host.
```

File: arch/arm/include/rp23xx/i2c_slave.h [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+/****************************************************************************
+ * Name: rp23xx_i2c0_slave_initialize
+ *
+ * Description:
+ *   Initialize I2C controller zero for slave operation, and return a pointer
+ *   to the instance of struct i2c_slave_s.  This function should only be
+ *   called once of a give controller.
+ *
+ *   Note: the same port cannot be initalized as both master and slave.
```
Comment (hartmannathan) [inline arch/arm/include/rp23xx/i2c_slave.h:58]:
```suggestion
 *   Note: the same port cannot be initialized as both master and slave.
```

File: arch/arm/include/rp23xx/i2c_slave.h [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+
+/****************************************************************************
+ * Name: rp23xx_i2c1_slave_initialize
+ *
+ * Description:
+ *   Initialize I2C controller zero for slave operation, and return a pointer
+ *   to the instance of struct i2c_slave_s.  This function should only be
+ *   called once of a give controller.
+ *
+ *   Note: the same port cannot be initalized as both master and slave.
```
Comment (hartmannathan) [inline arch/arm/include/rp23xx/i2c_slave.h:87]:
```suggestion
 *   Note: the same port cannot be initialized as both master and slave.
```

**Comment by acassis** [comment]
@PeterBee97 please take a look at this error:
```
arm-none-eabi-ld: warning: /github/workspace/sources/nuttx/nuttx has a LOAD segment with RWX permissions
arm-none-eabi-ld: /github/workspace/sources/nuttx/staging/libsched.a(nx_bringup.o): in function `nx_romfsetc':
/github/workspace/sources/nuttx/sched/init/nx_bringup.c:276:(.text.nx_bringup+0xa0): undefined reference to `romfs_img_len'
arm-none-eabi-ld: /github/workspace/sources/nuttx/sched/init/nx_bringup.c:276:(.text.nx_bringup+0xa4): undefined reference to `romfs_img'
make[1]: *** [Makefile:212: nuttx] Error 1
make: *** [tools/Unix.mk:551: nuttx] Error 2
make: Target 'all' not remade because of errors.
/github/workspace/sources/nuttx/tools/testbuild.sh: line 385: /github/workspace/sources/nuttx/../nuttx/nuttx.manifest: No such file or directory
  Normalize raspberrypi-pico-2/spisd
HEAD detached at pull/14831/merge
Untracked files:
  (use "git add <file>..." to include in what will be committed)
	boards/arm/rp23xx/raspberrypi-pico-2/scripts/memmap_default.ld.tmp
```

**Comment by PeterBee97** [comment]
> can we fuse the last three patches into the previous patches.

Sure. Squashed

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by acassis — APPROVED** [review]

**Comment by casaroli** [comment]
Oh nice, thank you!!

**Pull request merged**